### PR TITLE
Func.yaml V20180707 migration, init and deploy commands

### PR DESCRIPTION
**Pull request by JadeRedworth**
Running `fn migrate` will detect the version of the func.yaml (v1 or v2). 
- If v1 the contents of the file will be ran against the json schema to validate fields before migration takes place. Once passed a back up func file `func.yaml.bak` will be created. The newfuncFile object will be validated before the new `func.yaml` file is created.

- If the func.yaml is detected to be a v2 version, the user will be notified that they do not need to migrate.

`func.yaml` before:
```
name: jade-test
version: 0.0.6
runtime: go
entrypoint: ./func
type: sync
memory: 128
format: json
timeout: 120
path: /test
idle_timeout: 120
annotations:
 test:
  - 123
```

`func.yaml` after:
```
schema_version: 20180708
name: jade-test
version: 0.0.6
runtime: go
entrypoint: ./func
memory: 128
timeout: 120
idle_timeout: 120
annotations:
  test:
  - 123
triggers:
- name: jade-test
  type: http
  source: /jade-test
```

### CODE REVIEW

File: commands/migrate.go [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+	ff   *common.FuncFile
+	ffV2 *common.FuncFileV2
+}
+
+func MigrateCommand() cli.Command {
+	m := &migrateFnCmd{ff: &common.FuncFile{}, ffV2: &common.FuncFileV2{}}
+
+	return cli.Command{
+		Name:        "migrate",
+		Usage:       "Migrate a local func.yaml file to the new version",
```
Comment (rikgibson) [inline commands/migrate.go]: Probably would say "latest" rather than "new"

File: commands/migrate.go [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+		Category:    "DEVELOPMENT COMMANDS",
+		Aliases:     []string{"m"},
+		Description: "Migrate will detect the version of the current func.yaml file, to the new version.",
```
Comment (rikgibson) [inline commands/migrate.go]: "This command will detect the version of a func.yaml file and update it to match the latest version supported by the Fn CLI.  Any old or unsupported attributes will be removed, and any new ones may be added."

File: common/funcfile.go [hunk 3]
```diff
@@ -83,6 +83,37 @@ type FuncFile struct {
...
+	Config      map[string]string      `yaml:"config,omitempty" json:"config,omitempty"`
+	Annotations map[string]interface{} `yaml:"annotations,omitempty" json:"annotations,omitempty"`
+
+	Build []string `yaml:"build,omitempty" json:"build,omitempty"`
+
+	Expects  Expects   `yaml:"expects,omitempty" json:"expects,omitempty"`
+	Triggers []Trigger `yaml:"triggers,omitempty" json:"triggers,omitempty"`
+}
+
+// Trigger represents the
```
Comment (mantree) [inline common/funcfile.go]: typo

File: commands/migrate.go [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+		Name:        "migrate",
+		Usage:       "Migrate a local func.yaml file to the latest version",
+		Category:    "DEVELOPMENT COMMANDS",
+		Aliases:     []string{"m"},
+		Description: "This command will detect the version of a func.yaml file and update it to match the latest version supported by the Fn CLI. Any old or unsupported attributes will be removed, and any new ones may be added. The current func.yaml will be renamed to func.yaml.bak and a new func.yaml created",
+		Action:      m.migrate,
+	}
+}
+
+func (m *migrateFnCmd) migrate(c *cli.Context) error {
```
Comment (mantree) [inline commands/migrate.go]: As discussed, let's have the 'business logic' decisions up here.

File: common/funcfile.go [hunk 5]
```diff
@@ -83,6 +83,37 @@ type FuncFile struct {
 	Expects Expects `yaml:"expects,omitempty" json:"expects,omitempty"`
 }
 
+type FuncFileV2 struct {
```
Comment (mantree) [inline common/funcfile.go]: it's not v2 it's v20180707

File: common/schema.go [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+import (
+	"errors"
+	"fmt"
+	"path/filepath"
+
+	"github.com/xeipuuv/gojsonschema"
+)
+
+const schema = `{
+    "title": "V2 func file schema",
```
Comment (mantree) [inline common/schema.go]: not v2 etc etc

File: common/schema.go [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+                },
+                "source": {
+                    "type":"string"
+                }
+            }
+        }
+    }
+}`
+
+func ValidateSchema(jsonFile string) error {
```
Comment (mantree) [inline common/schema.go]: does this validate the schema? or does it actually validate the file?

Consider breaking this out so the schema is an arg.

Comment (JadeRedworth) [reply common/schema.go]: It validates the file against the schema. Or are you referring to the name of the function?

File: commands/migrate.go [hunk 8]
```diff
@@ -0,0 +1,177 @@
...
+	var err error
+	oldFF, err := readInFuncFile()
+	if err != nil {
+		return err
+	}
+
+	version := detectFuncYamlVersion(oldFF)
+	if version != latestYamlVersion {
```
Comment (mantree) [inline commands/migrate.go]: Shouldn't this be ==?

File: commands/migrate.go [hunk 9]
```diff
@@ -0,0 +1,177 @@
...
+	if err != nil {
+		return nil, err
+	}
+
+	return ff, nil
+}
+
+func detectFuncYamlVersion(oldFF map[string]interface{}) string {
+	if _, ok := oldFF["schema_version"]; !ok {
+		return latestYamlVersion
```
Comment (mantree) [inline commands/migrate.go]: I think this should return the actual value of "schema_version" if it exists, and "v1" if it doesn't.

These tend to be call 'getXXX'

File: commands/build.go [hunk 10]
```diff
@@ -56,6 +56,25 @@ func (b *buildcmd) build(c *cli.Context) error {
 
 	dir := common.GetDir(c)
 
+	ffV, err := common.ReadInFuncFile()
+	version := common.GetFuncYamlVersion(ffV)
+	if version == common.LatestYamlVersion {
```
Comment (mantree) [inline commands/build.go]: This would be better as a 'switch' statement.

switch version:
v2018: blah
v1: blah

File: commands/deploy.go [hunk 11]
```diff
@@ -162,14 +169,35 @@ func (p *deploycmd) deploySingle(c *cli.Context, appName string, appf *common.Ap
...
-
 	err := os.Chdir(dir)
 	if err != nil {
 		return err
 	}
 	defer os.Chdir(wd)
 
+	ffV, err := common.ReadInFuncFile()
+	version := common.GetFuncYamlVersion(ffV)
+	if version == common.LatestYamlVersion {
```
Comment (mantree) [inline commands/deploy.go]: again switch. The reason is to keep the function return statements close and understandable.

File: commands/deploy.go [hunk 12]
```diff
@@ -309,6 +332,49 @@ func (p *deploycmd) deployFunc(c *cli.Context, appName, baseDir, funcfilePath st
...
+func (p *deploycmd) deployFuncV20180707(c *cli.Context, appName, baseDir, funcfilePath string, funcfile *common.FuncFileV20180707) error {
+	if appName == "" {
+		return errors.New("App name must be provided, try `--app APP_NAME`")
+	}
+	//dir := filepath.Dir(funcfilePath)
+	// get name from directory if it's not defined
+	if funcfile.Name == "" {
+		funcfile.Name = filepath.Base(filepath.Dir(funcfilePath)) // todo: should probably make a copy of ff before changing it
+	}
+	// if funcfile.Path == "" {
```
Comment (mantree) [inline commands/deploy.go]: ditch this commented out stuff, or make it real.

File: commands/deploy.go [hunk 13]
```diff
@@ -320,6 +386,17 @@ func setRootFuncInfo(ff *common.FuncFile, appName string) {
 	}
 }
 
+func setFuncInfoV20180707(ff *common.FuncFileV20180707, appName string) {
+	if ff.Name == "" {
+		fmt.Println("Setting name")
+		ff.Name = fmt.Sprintf("%s-root", appName)
+	}
+	// if ff.en == "" {
```
Comment (mantree) [inline commands/deploy.go]: ditch or realise

File: commands/deploy.go [hunk 14]
```diff
@@ -320,6 +386,17 @@ func setRootFuncInfo(ff *common.FuncFile, appName string) {
 	}
 }
 
+func setFuncInfoV20180707(ff *common.FuncFileV20180707, appName string) {
```
Comment (mantree) [inline commands/deploy.go]: wow go is pant.s

File: commands/deploy.go [hunk 15]
```diff
@@ -329,6 +406,79 @@ func (p *deploycmd) updateRoute(c *cli.Context, appName string, ff *common.FuncF
...
+		}
+	} else {
+		fn.ID = fnRes.ID
+		err = function.PutFn(p.clientV2, fn.ID, fn)
+		if err != nil {
+			return err
+		}
+	}
+
+	if fnRes == nil {
```
Comment (mantree) [inline commands/deploy.go]: this looks a bit weird. why are we getting it again?

File: commands/deploy.go [hunk 16]
```diff
@@ -329,6 +406,79 @@ func (p *deploycmd) updateRoute(c *cli.Context, appName string, ff *common.FuncF
...
+			return err
+		}
+		app, err = apps.GetAppByName(appName)
+		if err != nil {
+			return err
+		}
+	}
+
+	fnRes, err := function.GetFnByName(p.clientV2, app.ID, ff.Name)
+	if err != nil {
```
Comment (mantree) [inline commands/deploy.go]: this seems too generic. If there's an error -> create? Can we be a bit more specific about the error that leads to a create? Not Found for instance.

File: commands/init.go [hunk 17]
```diff
@@ -115,15 +119,25 @@ func (a *initFnCmd) init(c *cli.Context) error {
 		dir = a.wd
 	}
 
+	if a.triggerType != "" {
```
Comment (mantree) [inline commands/init.go]: is this using the trigger type to detect version? Shouldn't we only be init'ing fns of the new version?

File: commands/push.go [hunk 18]
```diff
@@ -49,7 +49,29 @@ func (p *pushcmd) flags() []cli.Flag {
 // push the container, and finally it will update function's route. Optionally,
```
Comment (mantree) [inline commands/push.go]: comment out of date? mentions 'route'

**Pull request merged**